aten/src/ATen/native/GatedLinearUnit.cpp

```diff
@@ -86,5 +88,51 @@ Tensor glu_backward(const Tensor& grad_output, const Tensor& input, int64_t dim)
   return at::glu_backward_out(grad_input, grad_output, input, dim);
 }
 
+// HardGLU is a piecewise linear approximation
+Tensor& hardglu_out(Tensor& result, const Tensor& self, int64_t dim) {
+  SizesAndDim size_dim = get_sizes_and_wrap_dim(self, dim);
+  result.resize_(size_dim.first);
+  auto half_size = size_dim.first[size_dim.second];
+  TensorPair split_self = make_split_pair(self, half_size, size_dim.second);
+
+  auto iter = TensorIterator::binary_op(result, split_self.first,
+                                        split_self.second);
+  hardglu_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor hardglu(const Tensor& self, int64_t dim) {
+  auto result = at::empty({0}, self.options());
+  return at::hardglu_out(result, self, dim);
+}
+
+Tensor& hardglu_backward_out(Tensor& grad_input,
```

Comment:
Not sure if I follow the math here. The gradient for hard GLU has two components:
HGLU(x) = x[0:n-1].Hsigmoid([x[n:end]) 
gradient[i] = HS[i+n]  * y[i] if i <n
 else: 
gradient[i+N] = x[0:n-1].Gradient(Hsigmoid). 
Is this what is being done by the code below?